### PR TITLE
SHARD-1058: limits on socket connections, subscriptions per socket, connection timeout +fixes

### PR DESCRIPTION
**Pull request by S0naliThakur**
Linear: https://linear.app/shm/issue/SHARD-1058/
Summary:
This PR introduces the following changes:

1. Added limits on open socket connections, subscriptions per socket, and a one-day timeout for connections.
2. Enabled automatic closure of socket connections without active subscribers.
3. Fixed issues in the ratelimit middleware unit test.

### CODE REVIEW

File: src/websocket/index.ts [hunk 1]
```diff
@@ -21,7 +21,22 @@ interface Request {
   params: Params
 }
 
+// Add connection counter
+let activeConnections = 0
+
 export const onConnection = async (socket: WebSocket.WebSocket): Promise<void> => {
+  // Check max connections limit
+  if (activeConnections >= CONFIG.websocket.maxConnections) {
+    socket.close(1013, 'Maximum number of connections reached')
```
Comment (aniketdivekar) [inline src/websocket/index.ts]: Should we say `server busy, please try again after some time` ?

File: test/unit/middlewares/rateLimit.test.ts [hunk 2]
```diff
@@ -42,13 +42,15 @@ describe('Rate Limiting', () => {
...
+    // Mock console.error
+    jest.spyOn(console, 'error').mockImplementation(() => {})
   })
 
   it('should skip rate limiting when disabled in config', async () => {
     CONFIG.rateLimit = false
 
     const mockReq = {
-      socket: { remoteAddress: '127.0.0.1' },
+      ip: '127.0.0.1',
```
Comment (aniketdivekar) [inline test/unit/middlewares/rateLimit.test.ts:53]: why is this impacted w/ these changes?

Comment (S0naliThakur) [reply test/unit/middlewares/rateLimit.test.ts:53]: the unit testing fails as the flow expects ip in the request post changes as per this [comment](https://github.com/shardeum/json-rpc-server/pull/91#discussion_r1862487043)

File: src/websocket/index.ts (hunk 1 above)
Comment (S0naliThakur) [reply src/websocket/index.ts]: try again later is closer to code 1003. Shall we change the code too>

Comment (aniketdivekar) [reply src/websocket/index.ts]: yes

Comment (S0naliThakur) [reply src/websocket/index.ts]: done

**Review by abdulazeem-tk4vr — APPROVED** [review]

**Review by aniketdivekar — APPROVED** [review]

File: src/websocket/index.ts [hunk 3]
```diff
@@ -21,7 +21,22 @@ interface Request {
   params: Params
 }
 
+// Add connection counter
+let activeConnections = 0
+
 export const onConnection = async (socket: WebSocket.WebSocket): Promise<void> => {
+  // Check max connections limit
+  if (activeConnections >= CONFIG.websocket.maxConnections) {
+    socket.close(1003, 'Server busy. Please try again later.')
```
Comment (BelfordZ) [inline src/websocket/index.ts:30]: will this properly clean up all the associated subscriptions?

File: test/unit/websocket/index.test.ts [hunk 4]
```diff
@@ -0,0 +1,226 @@
+import WebSocket from 'ws'
```
Comment (BelfordZ) [inline test/unit/websocket/index.test.ts:1]: image

**Review by BelfordZ — APPROVED** [review]
great work.

Further improvements: in the timeout, lets reduce the timer, and make it check if the socket is 'active' aka lets track the usage of the socket, and only drop the ones that are inactive, until we hit a hard limit.

we can do that next time though. LGTM

File: src/websocket/index.ts (hunk 3 above)
Comment (S0naliThakur) [reply src/websocket/index.ts:30]: Here we are rejecting the connection before any subscriptions could be created, so yes that's handled :)

**Pull request merged**